components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/dao/impl/AuthenticatorManagementDAOImpl.java

```diff
@@ -0,0 +1,276 @@
+import org.wso2.carbon.database.utils.jdbc.exceptions.TransactionException;
+import org.wso2.carbon.identity.application.common.constant.AuthenticatorMgtSQLConstants.Column;
+import org.wso2.carbon.identity.application.common.constant.AuthenticatorMgtSQLConstants.Query;
+import org.wso2.carbon.identity.application.common.dao.AuthenticatorManagementDAO;
+import org.wso2.carbon.identity.application.common.exception.AuthenticatorMgtException;
+import org.wso2.carbon.identity.application.common.exception.AuthenticatorMgtServerException;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.application.common.model.UserDefinedLocalAuthenticatorConfig;
+import org.wso2.carbon.identity.application.common.util.AuthenticatorMgtExceptionBuilder.AuthenticatorMgtError;
+import org.wso2.carbon.identity.base.AuthenticatorPropertyConstants.AuthenticationType;
+import org.wso2.carbon.identity.base.AuthenticatorPropertyConstants.DefinedByType;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.application.common.util.AuthenticatorMgtExceptionBuilder.buildServerException;
+
+/**
+ * This class implements the AuthenticatorManagementDAO interface which perform CRUD operation on database.
+ */
+public class AuthenticatorManagementDAOImpl implements AuthenticatorManagementDAO {
+
+    public static final String IS_TRUE_VALUE = "1";
+    public static final String IS_FALSE_VALUE = "0";
+    public static final String LOCAL_IDP_NAME = "LOCAL";
+
+    @Override
+    public UserDefinedLocalAuthenticatorConfig addUserDefinedLocalAuthenticator(
+            UserDefinedLocalAuthenticatorConfig authenticatorConfig, int tenantId) throws AuthenticatorMgtException {
+
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        try {
+            jdbcTemplate.withTransaction(template ->
+                jdbcTemplate.executeInsert(Query.ADD_AUTHENTICATOR_SQL,
+                    (statement -> {
+                        statement.setString(Column.NAME, authenticatorConfig.getName());
+                        statement.setString(Column.DISPLAY_NAME, authenticatorConfig.getDisplayName());
+                        statement.setString(Column.DEFINED_BY, authenticatorConfig.getDefinedByType().toString());
+                        statement.setString(Column.AUTHENTICATION_TYPE, authenticatorConfig.getAuthenticationType()
+                                .toString());
+                        statement.setInt(Column.IS_ENABLED, authenticatorConfig.isEnabled() ? 1 : 0);
+                        statement.setString(Column.IDP_NAME, LOCAL_IDP_NAME);
+                        statement.setInt(Column.TENANT_ID, tenantId);
+                    }), null, false));
+
+            int authenticatorConfigID = getAuthenticatorEntryId(authenticatorConfig.getName(), tenantId);
```

Comment:
We can get the auto generated id from the withTransaction() method by passing the third argument as true. 
If that returned value is zero, then we can use this query to fetch the id. 

Invoking this conditionally may reduce an additional db query in most of the time as most jdbc drivers supports returning the auto generated id